File: frontend/src/services/Libro/LibroService.ts
import HttpClient from './../HttpClient';
import LibroGetAllResponseDto from '../../dtos/Libro/getAll/LibroGetAllResponseDto';
import LibroGetByIdResponseDto from '../../dtos/Libro/getById/LibroGetByIdResponseDto';
import LibroGetAllForAssociationResponseDto from '../../dtos/Libro/getAllForAssociation/LibroGetAllForAssociationResponseDto';
import LibroCreateRequestDto from '../../dtos/Libro/create/LibroCreateRequestDto';

class LibroService
{
   getAllLibro = async (): Promise<LibroGetAllResponseDto[]> => {
       const response = await HttpClient.get(`api/v1/Libro/`);
       const bodyReponse = await response.json();
       return bodyReponse as LibroGetAllResponseDto[];
   };
   getAllForAssociationLibro = async (): Promise<LibroGetAllForAssociationResponseDto[]> => {
       const response = await HttpClient.get(`api/v1/Libro/getAllForAssociations`);
       const bodyReponse = await response.json();
       return bodyReponse as LibroGetAllForAssociationResponseDto[];
   };

    getById = async (id:number): Promise<LibroGetByIdResponseDto> => {
       const response = await HttpClient.get(`api/v1/Libro/${id}/`);
       const bodyResponse = await response.json();
       return bodyResponse as LibroGetByIdResponseDto;
    }

   postCreateLibro = async (libroCreateRequestDto : LibroCreateRequestDto): Promise<Response> => {
       return await HttpClient.post(`api/v1/Libro/`, libroCreateRequestDto);
   };

   deleteLibro = async (id : number): Promise<Response> => {
       return await HttpClient.delete(`api/v1/Libro/${id}`);
   };


}

export default new LibroService()

File: backend/src/main/java/com/example/template/services/LibroService.java
package com.example.template.services;

import com.example.template.dtos.*;
import com.example.template.exceptions.LibroDontFoundException;
import com.example.template.model.*;
import com.example.template.repositories.*;
import com.example.template.services.ILibroService;
import com.example.template.mappers.*;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class LibroService implements ILibroService {

    protected LibroRepository LibroRepository;
    protected AutorRepository autorRepository;
    protected MapperLibro mapperLibro;


        @Override
    public void create(LibroCreateRequestDTO libroCreateRequestDTO) {
       Libro libro;
       libro = mapperLibro.CreateDtoToLibro(libroCreateRequestDTO);
       
	   if(libroCreateRequestDTO.getAutorId() != 0) {
	       Optional<Autor> optionalAutor = autorRepository.findById(libroCreateRequestDTO.getAutorId());
	
	       if(optionalAutor.isPresent()){
	       libro.setAutor(optionalAutor.get());
	       }
	    }
       LibroRepository.save(libro);
    }
    @Override
    public LibroGetByIdResponseDTO getById(long id) throws LibroDontFoundException {
        Optional<Libro> optional = LibroRepository.findById(id);
        if (optional.isEmpty()){
            throw new LibroDontFoundException(id);
        }
        Libro libro = optional.get();


        LibroGetByIdResponseDTO libroGetByIdResponseDTO = mapperLibro.toLibroGetByIdResponseDTO(libro);

        return libroGetByIdResponseDTO;
    }

    @Override
    public List<LibroResponseDTO> getAll() {
        List<Libro> LibroList;
        List<LibroResponseDTO> LibroResponseDTOList = new ArrayList<>();

        LibroList = LibroRepository.findAll();

        for (Libro Libro: LibroList) {
                LibroResponseDTO LibroResponseDTO = mapperLibro.toLibroResponseDTO(Libro);
                LibroResponseDTOList.add(LibroResponseDTO);
        }
        return LibroResponseDTOList;
    }
    @Override
    public List<LibroGetAllForAssociationResponseDTO> getAllForAssociation() {
        List<Libro> LibroList;
        List<LibroGetAllForAssociationResponseDTO> LibroGetAllForAssociationResponseDTOList = new ArrayList<>();

        LibroList = LibroRepository.findAll();

        for (Libro Libro: LibroList) {
                LibroGetAllForAssociationResponseDTO LibroGetAllForAssociationResponseDTO = mapperLibro.toLibroGetAllForAssociationResponseDTO(Libro);
                LibroGetAllForAssociationResponseDTOList.add(LibroGetAllForAssociationResponseDTO);
        }
        return LibroGetAllForAssociationResponseDTOList;
    }

    @Override
    public LibroResponseDTO update(long id, LibroRequestDTO LibroRequestDTO) throws LibroDontFoundException {
        Optional<Libro> LibroOptional  = LibroRepository.findById(id);
        if (LibroOptional.isEmpty()){
            throw new LibroDontFoundException(id);
        }
        Libro Libro = LibroOptional.get();

        Libro LibroUpdate = mapperLibro.toLibro(LibroRequestDTO,id);

       LibroUpdate = LibroRepository.save(LibroUpdate);

        LibroResponseDTO LibroResponseDTO = mapperLibro.toLibroResponseDTO(LibroUpdate);

        return LibroResponseDTO;
    }

    @Override
    public boolean delete(long id) throws LibroDontFoundException {
        Optional<Libro> LibroOptional  = LibroRepository.findById(id);

        if (LibroOptional.isEmpty()){
            throw new LibroDontFoundException(id);
        }

        Libro Libro = LibroOptional.get();
        LibroRepository.delete(Libro);

        return true;
    }
}
